taf/utils.py

```diff
@@ -33,6 +36,84 @@ def convert(self, value, param, ctx):
 ISO_DATE_PARAM_TYPE = IsoDateParamType()
 
 
+def extract_x509(cert_pem):
+  from cryptography import x509
+  from cryptography.hazmat.backends import default_backend
+
+  cert = x509.load_pem_x509_certificate(cert_pem, default_backend())
+
+  def _get_attr(oid):
+    attrs = cert.subject.get_attributes_for_oid(oid)
+    return attrs[0].value if len(attrs) > 0 else ""
+
+  return {
+      "name": _get_attr(x509.OID_COMMON_NAME),
+      "organization": _get_attr(x509.OID_ORGANIZATION_NAME),
+      "country": _get_attr(x509.OID_COUNTRY_NAME),
+      "state": _get_attr(x509.OID_STATE_OR_PROVINCE_NAME),
+      "locality": _get_attr(x509.OID_LOCALITY_NAME),
+      "valid_from": cert.not_valid_before.strftime("%Y-%m-%d"),
+      "valid_to": cert.not_valid_after.strftime("%Y-%m-%d"),
+  }
+
+
+def get_cert_names_from_keyids(certs_dir, keyids):
+  cert_names = []
+  for keyid in keyids:
+    try:
+      name = extract_x509((Path(certs_dir) / keyid + ".pem").read_bytes())['name']
+      if not name:
+        print("Cannot extract common name from x509, using key id instead.")
+        cert_names.append(keyid)
+      else:
+        cert_names.append(name)
+    except FileNotFoundError:
+      print("Certificate does not exist ({}).".format(keyid))
+  return cert_names
+
+
+def get_pin_for(name, confirm=True, repeat=True):
+  pin = getpass('Enter PIN for {}: '.format(name))
+  if confirm:
+    if pin != getpass('Confirm PIN for {}: '.format(name)):
+      err_msg = "PINs doesn't match!"
+      if repeat:
+        print(err_msg)
+        get_pin_for(name, confirm, repeat)
+      else:
+        raise PINMissmatchError(err_msg)
+  return pin
+
+
+def get_yubikey_pin_for_keyid(expected_keyids, key_slot=(2,), holders_name=' '):
+  from taf.repository_tool import get_yubikey_public_key
+  from oll_sc.api import sc_is_present
+  from oll_sc.exceptions import SmartCardError
+
+  if isinstance(expected_keyids, str):
+    expected_keyids = (expected_keyids)
+  if isinstance(key_slot, int) or isinstance(key_slot, str):
+    key_slot = (int(key_slot),)
+
+  while True:
+    try:
+      input("Please insert {}'s yubikey and press ENTER.\n".format(holders_name))
+      if not sc_is_present():
+        continue
+
+      key_pin = get_pin_for(holders_name, confirm=False)
+      inserted_key = get_yubikey_public_key(key_slot, key_pin)
+      if inserted_key['keyid'] not in expected_keyids:
+        print("Please insert valid yubikey!")
```

Comment:
When setting up repository, if a user's inserts root1 Yubikey instead of root2 yubikey when prompted to do so, he will be asked to enter a pin and then asked to insert a valid Yubikey.
If that's not hard to do, can we check if the key matches the expected one before a user is asked to insert the PIN? Also, valid might not be the best word here, since the inserted Yubikey is valid, but not the one we expect

Reply:
It is possible, but I would need to change `oll-sc`. Since we want to get rid of `oll-sc` and move existing yubikey functions to TAF, I will address this comment after that work.